recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py

```diff
@@ -0,0 +1,239 @@
+            'build': self.build
+        }
+
+fs_manifest_format = '%p\t%M\t%u\t%g\t%l'
+fs_manifest_columns = ['path', 'mode', 'user', 'group', 'link_target']
+def get_fs_manifest():
+    search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
+    omit_dirs = ['/lib/modules', '/var/cache', '/var/run', '/var/tmp', '/var/volatile']
+
+    omit_expr = []
+    for d in omit_dirs:
+        omit_expr += ['-path', d, '-o']
+    omit_expr.pop()
+
+    fs_manifest = run_cmd(
+        ['find']
+        + search_dirs
+        + ['-printf', fs_manifest_format + '\n']
+        + ['('] + omit_expr + [')', '-prune']
+    )
+    return fs_manifest
+
+def upload_manifest(db, fs_manifest, logger):
+    data = {}
+    os_version = OsVersion()
+    data['os_version_codename'] = os_version.codename
+    data['os_version_full'] = os_version.full
+    data['os_version_major_minor'] = os_version.major_minor
+    data['os_version'] = os_version.version_dict
+    data['date'] = str(datetime.datetime.now())
+
+    header = ''
+    for key, val in data.items():
+        header = header + '# {}: {}\n'.format(key, val)
+
+    data['fs_permissions'] = header + fs_manifest
+
+    record = db.insert(data)
+    logger.log('INFO: Uploaded fs permissions of OS {} from {} with _id {}'.format(data['os_version_full'], data['date'], record.inserted_id))
+
+def strip_headers(fs_manifest):
+    return ''.join(line + '\n' for line in fs_manifest.splitlines() if not line.startswith('#'))
+
+def get_old_fs_manifest(db, logger):
+    os_version = OsVersion()
+    query = {
+        'os_version_codename': os_version.codename,
+        'os_version_major_minor': os_version.major_minor,
+        'os_version': {
+            '$lt': {
+                'major': os_version.major,
+                'minor': os_version.minor,
+                'patch': os_version.patch,
+                # $lt considers booleans to be greater than all numbers
+                'phase': False,
+                'build': False
+            }
+        },
+        'os_version.phase': OsVersion.phase_number['f']
+    }
+
+    if db.count_documents(query):
+        results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+    else:
+        # If there are no results, there may not be a run from the current os version.
+        # Remove that requirement.
+        del query['os_version_major_minor']
+        logger.log('INFO: No prior fs permissions from this OS Version found. Using previous version.')
+
+        if db.count_documents(query):
+            results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+        else:
+            # Keep this log in first line to help streak indexer group results. Hashes will be populated at end.
+            logger.first_log('INFO: fs_permissions_diff: {} against <empty> '.format(os_version.full))
+
+            logger.log('INFO: No suitable previous fs permissions found')
+            return ''
```

Comment:
Conversation with @jpautler on improving output for readers of diff output.

 - Consider having a double-diff (analogue: git's three-way merge) against a basis (version < maj.min.pat.0) and recent (version < maj.min.pat.bld) to consistently report differences, but also point out new ones.
  - Attempt to match codename if possible, but if not possible, relax that constraint (and assume there's only one alternative, or equivalently. highest version found will be in the "previous" codename)
  - Try sorting by full version instead of date

```suggestion
   os_version = OsVersion()
    query = {
        'os_version_codename': os_version.codename,
        'os_version': {
            '$lt': {
                'major': os_version.major,
                'minor': os_version.minor,
                'patch': os_version.patch,
                'build': 0
            }
        }
    }

    if db.count_documents(query):
        # TODO sort change
        results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
    else:
        # If there are no results, there may not be a run from the current os codename.
        # Remove that requirement.
        del query['os_version_codename']
        logger.log('INFO: No prior fs permissions from this codename found. Relaxing that constraint.')

        if db.count_documents(query):
            # TODO sort change
            results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
        else:
            # Keep this log in first line to help streak indexer group results. Hashes will be populated at end.
            logger.first_log('INFO: fs_permissions_diff: {} against <empty> '.format(os_version.full))

            logger.log('INFO: No suitable previous fs permissions found')
            return ''
```

Also make sure that `os_version.phase` isn't uploaded to MongoDB. (Do this before enabling event so that $lt doesn't have undesirable behavior with key count stuff.)

Reply:
Possible printout representation: similar to git's 3-way merge with two columns of +/-/space